pkg/repository/v1_repository.go

```diff
@@ -54,6 +65,87 @@ func NewV1Repo(mirror Mirror, opts Options, local v1manifest.LocalManifests) *V1
 const maxTimeStampSize uint = 1024
 const maxRootSize uint = 1024 * 1024
 
+// UpdateComponents updates the components described by specs.
+// TODO tests
+func (r *V1Repository) UpdateComponents(specs []ComponentSpec) error {
+	err := r.ensureManifests()
+	if err != nil {
+		return err
+	}
+	for _, spec := range specs {
+		manifest, err := r.updateComponentManifest(spec.ID)
+		if err != nil {
+			// TODO could report error and continue
+			return err
+		}
+
+		platform := utils.PlatformString()
+		versions, ok := manifest.Platforms[platform]
+		if !ok {
+			// TODO could report error and continue
+			return fmt.Errorf("platform %s not supported by component %s", platform, spec.ID)
+		}
+
+		version, versionItem, err := r.selectVersion(spec.ID, versions, spec.Version)
+		if err != nil {
+			// TODO could report error and continue
+			return err
+		}
+
+		if !spec.Force {
+			installed, err := r.local.ComponentInstalled(spec.ID, version)
+			if err != nil {
+				return err
+			}
+			if installed {
+				// TODO could report error and continue
+				return fmt.Errorf("component %s version %s is already installed", spec.ID, version)
+			}
+		}
+
+		reader, err := r.downloadComponent(versionItem)
+		if err != nil {
+			// TODO could report error and continue
+			return err
+		}
+
+		err = r.local.InstallComponent(reader, spec.ID, version)
+		if err != nil {
+			// TODO could report error and continue
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *V1Repository) ensureManifests() error {
+	// TODO
```

Comment:
this will ensure we can the fresh and right snapshot&index?

Reply:
It will ensure that we have an up to date index, root, and snapshot, i.e., that we can call updateComponentManifest